File: android/src/main/java/com/margelo/nitro/nitrotimer/NitroTimer.kt
package com.margelo.nitro.nitrotimer

import android.graphics.Color
import android.graphics.Typeface
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.core.graphics.toColorInt
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.roundToInt

@DoNotStrip
class HybridNitroTimer(val context: ThemedReactContext) : HybridNitroTimerSpec() {

    // View with TextView for timer display
    override val view: FrameLayout = FrameLayout(context)
    private val timerTextView = TextView(context)

    // Timer properties
    private var endDate: Date? = null
    private var isRunning = false
    private var isPaused = false
    private var wasExpired = false
    private var wasCritical = false
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.getDefault()).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    // Props
    private var _endTime = ""
    override var endTime: String
        get() = _endTime
        set(value) {
            _endTime = value
            setEndTimeFromString(value)
        }

    private var _autoStart = true
    override var autoStart: Boolean
        get() = _autoStart
        set(value) {
            _autoStart = value
            if (value && !isRunning) {
                start()
            }
        }

    private var _criticalThreshold = 300.0 // 5 minutes default
    override var criticalThreshold: Double
        get() = _criticalThreshold
        set(value) {
            _criticalThreshold = value
        }

    // Display format
    private var _format = "HH:MM:SS"
    override var format: String
        get() = _format
        set(value) {
            _format = value
        }

    private var _showDays = true
    override var showDays: Boolean
        get() = _showDays
        set(value) {
            _showDays = value
        }

    private var _hideZeroHours = false
    override var hideZeroHours: Boolean
        get() = _hideZeroHours
        set(value) {
            _hideZeroHours = value
        }

    // Styling
    private var _fontSize = 16.0
    override var fontSize: Double
        get() = _fontSize
        set(value) {
            _fontSize = value
            updateTextViewStyle()
        }

    private var _textColor = "#000000"
    override var textColor: String
        get() = _textColor
        set(value) {
            _textColor = value
            updateTextViewStyle()
        }

    private var _criticalColor = "#FF0000"
    override var criticalColor: String
        get() = _criticalColor
        set(value) {
            _criticalColor = value
        }

    private var _fontWeight = "normal"
    override var fontWeight: String
        get() = _fontWeight
        set(value) {
            _fontWeight = value
            updateTextViewStyle()
        }

    private var _textAlign = "left"
    override var textAlign: String
        get() = _textAlign
        set(value) {
            _textAlign = value
            updateTextViewStyle()
        }

    private var _backgroundColor: String? = null
    override var backgroundColor: String?
        get() = _backgroundColor
        set(value) {
            _backgroundColor = value
            updateViewStyle()
        }

    private var _cornerRadius: Double? = null
    override var cornerRadius: Double?
        get() = _cornerRadius
        set(value) {
            _cornerRadius = value
            updateViewStyle()
        }

    private var _padding: Double? = null
    override var padding: Double?
        get() = _padding
        set(value) {
            _padding = value
            updateTextViewLayout()
        }

    // Behavior
    private var _pauseWhenNotVisible = false
    override var pauseWhenNotVisible: Boolean
        get() = _pauseWhenNotVisible
        set(value) {
            _pauseWhenNotVisible = value
        }

    private var _continueInBackground = true
    override var continueInBackground: Boolean
        get() = _continueInBackground
        set(value) {
            _continueInBackground = value
        }

    // Events
    override var onExpired: (() -> Unit)? = null
    override var onCritical: (() -> Unit)? = null

    init {
        setupTimerTextView()
        TimerManager.addTimer(this)
    }

    fun destroy() {
        TimerManager.removeTimer(this)
    }

    private fun setupTimerTextView() {
        view.addView(timerTextView)
        updateTextViewLayout()
        updateTextViewStyle()
        updateViewStyle()

        timerTextView.text = formatTime(0, 0, 0, 0)
    }

    private fun updateTextViewLayout() {
        val paddingPx = (_padding ?: 0.0).dpToPx().roundToInt()
        val layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
        layoutParams.setMargins(paddingPx, paddingPx, paddingPx, paddingPx)
        timerTextView.layoutParams = layoutParams
        timerTextView.setPadding(paddingPx, paddingPx, paddingPx, paddingPx)
    }

    private fun updateTextViewStyle() {
        timerTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, _fontSize.toFloat())
        timerTextView.setTextColor(_textColor.toColorInt())

        timerTextView.typeface = if (_fontWeight == "bold") {
            Typeface.DEFAULT_BOLD
        } else {
            Typeface.DEFAULT
        }

        timerTextView.gravity = when (_textAlign) {
            "center" -> Gravity.CENTER
            "right" -> Gravity.END or Gravity.CENTER_VERTICAL
            else -> Gravity.START or Gravity.CENTER_VERTICAL
        }
    }

    private fun updateViewStyle() {
        _backgroundColor?.let { bgColor ->
            view.setBackgroundColor(bgColor.toColorInt())
        }

        _cornerRadius?.let { radius ->
            view.background?.let { drawable ->
                // Note: For proper corner radius, you'd want to use a GradientDrawable
                // This is a simplified version
            }
        }
    }

    private fun setEndTimeFromString(timeString: String) {
        try {
            endDate = dateFormat.parse(timeString)
            updateDisplay()
        } catch (e: Exception) {
            // Handle parsing error
            endDate = null
        }
    }

    fun updateDisplay() {
        val endDate = this.endDate ?: run {
            timerTextView.text = formatTime(0, 0, 0, 0)
            return
        }

        val now = Date()
        val timeInterval = (endDate.time - now.time) / 1000.0

        if (timeInterval <= 0) {
            // Timer expired
            timerTextView.text = formatTime(0, 0, 0, 0)
            if (!wasExpired) {
                wasExpired = true
                onExpired?.invoke()
            }
            return
        }

        wasExpired = false

        // Check critical threshold
        val isCritical = timeInterval <= _criticalThreshold
        if (isCritical && !wasCritical) {
            wasCritical = true
            onCritical?.invoke()
        } else if (!isCritical) {
            wasCritical = false
        }

        // Update text color based on critical state
        val currentColor = if (isCritical) _criticalColor else _textColor
        timerTextView.setTextColor(currentColor.toColorInt())

        // Calculate time components
        val totalSeconds = timeInterval.toInt()
        val days = totalSeconds / 86400
        val hours = (totalSeconds % 86400) / 3600
        val minutes = (totalSeconds % 3600) / 60
        val seconds = totalSeconds % 60

        timerTextView.text = formatTime(days, hours, minutes, seconds)
    }

    private fun formatTime(days: Int, hours: Int, minutes: Int, seconds: Int): String {
        return when (_format) {
            "HH:MM:SS" -> {
                when {
                    _showDays && days > 0 -> String.format("%d:%02d:%02d:%02d", days, hours, minutes, seconds)
                    _hideZeroHours && hours == 0 -> String.format("%02d:%02d", minutes, seconds)
                    else -> String.format("%02d:%02d:%02d", hours, minutes, seconds)
                }
            }
            "H:M:S" -> {
                when {
                    _showDays && days > 0 -> String.format("%d:%d:%d:%d", days, hours, minutes, seconds)
                    _hideZeroHours && hours == 0 -> String.format("%d:%d", minutes, seconds)
                    else -> String.format("%d:%d:%d", hours, minutes, seconds)
                }
            }
            "Hh Mm Ss" -> {
                when {
                    _showDays && days > 0 -> String.format("%dd %dh %dm %ds", days, hours, minutes, seconds)
                    _hideZeroHours && hours == 0 -> String.format("%dm %ds", minutes, seconds)
                    else -> String.format("%dh %dm %ds", hours, minutes, seconds)
                }
            }
            "H hours M minutes" -> {
                when {
                    _showDays && days > 0 -> String.format("%d days %d hours %d minutes", days, hours, minutes)
                    _hideZeroHours && hours == 0 -> String.format("%d minutes", minutes)
                    else -> String.format("%d hours %d minutes", hours, minutes)
                }
            }
            else -> String.format("%02d:%02d:%02d", hours, minutes, seconds)
        }
    }

    // MARK: - Timer Methods

    override fun start() {
        isRunning = true
        isPaused = false
        TimerManager.startTimer()
    }

    override fun pause() {
        isPaused = true
    }

    override fun resume() {
        isPaused = false
    }

    override fun reset() {
        wasExpired = false
        wasCritical = false
        updateDisplay()
    }

    override fun addSeconds(seconds: Double) {
        endDate?.let { currentEndDate ->
            endDate = Date(currentEndDate.time + (seconds * 1000).toLong())
            updateDisplay()
        }
    }

    override fun setEndTime(endTime: String) {
        setEndTimeFromString(endTime)
    }

    override fun getCurrentTime(): Map<String, Any> {
        val endDate = this.endDate ?: return mapOf(
            "days" to 0,
            "hours" to 0,
            "minutes" to 0,
            "seconds" to 0,
            "totalSeconds" to 0,
            "isExpired" to true,
            "isCritical" to false
        )

        val now = Date()
        val timeInterval = (endDate.time - now.time) / 1000.0

        if (timeInterval <= 0) {
            return mapOf(
                "days" to 0,
                "hours" to 0,
                "minutes" to 0,
                "seconds" to 0,
                "totalSeconds" to 0,
                "isExpired" to true,
                "isCritical" to false
            )
        }

        val totalSeconds = timeInterval.toInt()
        val days = totalSeconds / 86400
        val hours = (totalSeconds % 86400) / 3600
        val minutes = (totalSeconds % 3600) / 60
        val seconds = totalSeconds % 60

        return mapOf(
            "days" to days,
            "hours" to hours,
            "minutes" to minutes,
            "seconds" to seconds,
            "totalSeconds" to totalSeconds,
            "isExpired" to false,
            "isCritical" to timeInterval <= _criticalThreshold
        )
    }

    private fun Double.dpToPx(): Double {
        val density = context.resources.displayMetrics.density
        return this * density
    }
}

// MARK: - Timer Manager Singleton

object TimerManager {
    private val handler = Handler(Looper.getMainLooper())
    private val timers = mutableListOf<HybridNitroTimer>()
    private var updateRunnable: Runnable? = null
    private var isRunning = false
    private var isAppInBackground = false

    fun addTimer(timer: HybridNitroTimer) {
        timers.add(timer)
    }

    fun removeTimer(timer: HybridNitroTimer) {
        timers.remove(timer)
        if (timers.isEmpty()) {
            stopTimer()
        }
    }

    fun startTimer() {
        if (!isRunning) {
            isRunning = true
            scheduleUpdate()
        }
    }

    fun onAppEnterBackground() {
        isAppInBackground = true
        // Pause timers that shouldn't continue in background
        for (timer in timers.toList()) {
            if (!timer._continueInBackground) {
                timer.pause()
            }
        }
    }

    fun onAppEnterForeground() {
        isAppInBackground = false
        // Resume timers that were paused due to backgrounding
        for (timer in timers.toList()) {
            if (!timer._continueInBackground && timer._autoStart) {
                timer.resume()
            }
        }
        // Force update all timers for accuracy
        updateTimers()
    }

    private fun stopTimer() {
        isRunning = false
        updateRunnable?.let { handler.removeCallbacks(it) }
        updateRunnable = null
    }

    private fun scheduleUpdate() {
        updateRunnable?.let { handler.removeCallbacks(it) }

        updateRunnable = Runnable {
            updateTimers()
            if (isRunning) {
                scheduleUpdate()
            }
        }

        handler.postDelayed(updateRunnable!!, 16) // ~60 FPS
    }

    private fun updateTimers() {
        for (timer in timers.toList()) {
            val shouldUpdate = timer.isRunning && !timer.isPaused

            // Handle visibility-based pausing
            if (timer._pauseWhenNotVisible) {
                // In a real implementation, you'd check if the timer's view is visible
                // For now, we'll assume timers are visible when app is in foreground
                val isVisible = !isAppInBackground
                if (!isVisible && shouldUpdate) {
                    continue // Skip update for invisible timers
                }
            }

            if (shouldUpdate) {
                timer.updateDisplay()
            }
        }
    }
}

File: ios/NitroTimer.swift
import UIKit
import Foundation

class HybridNitroTimer : HybridNitroTimerSpec {

    // UIView with UILabel for timer display
    var view: UIView = UIView()
    private let timerLabel = UILabel()

    // Timer properties
    private var endDate: Date?
    private var isRunning = false
    private var isPaused = false
    private var wasExpired = false
    private var wasCritical = false
    private let dateFormatter = ISO8601DateFormatter()

    // Props
    var endTime: String = "" {
        didSet {
            setEndTimeFromString(endTime)
        }
    }

    var autoStart: Bool = true {
        didSet {
            if autoStart && !isRunning {
                start()
            }
        }
    }

    var criticalThreshold: Double = 300.0 // 5 minutes default

    // Display format
    var format: String = "HH:MM:SS"
    var showDays: Bool = true
    var hideZeroHours: Bool = false

    // Styling
    var fontSize: Double = 16.0 {
        didSet {
            updateLabelStyle()
        }
    }

    var textColor: String = "#000000" {
        didSet {
            updateLabelStyle()
        }
    }

    var criticalColor: String = "#FF0000"

    var fontWeight: String = "normal" {
        didSet {
            updateLabelStyle()
        }
    }

    var textAlign: String = "left" {
        didSet {
            updateLabelStyle()
        }
    }

    var backgroundColor: String? = nil {
        didSet {
            updateViewStyle()
        }
    }

    var cornerRadius: Double? = nil {
        didSet {
            updateViewStyle()
        }
    }

    var padding: Double? = nil {
        didSet {
            updateLabelConstraints()
        }
    }

    // Behavior
    var pauseWhenNotVisible: Bool = false
    var continueInBackground: Bool = true

    // Events
    var onExpired: (() -> Void)?
    var onCritical: (() -> Void)?

    override init() {
        super.init()
        setupTimerLabel()
        TimerManager.shared.addTimer(self)
    }

    deinit {
        TimerManager.shared.removeTimer(self)
    }

    private func setupTimerLabel() {
        view.addSubview(timerLabel)
        timerLabel.translatesAutoresizingMaskIntoConstraints = false
        updateLabelConstraints()
        updateLabelStyle()
        updateViewStyle()

        timerLabel.text = formatTime(days: 0, hours: 0, minutes: 0, seconds: 0)
    }

    private func updateLabelConstraints() {
        timerLabel.removeFromSuperview()
        view.addSubview(timerLabel)
        timerLabel.translatesAutoresizingMaskIntoConstraints = false

        let paddingValue = padding ?? 0.0
        NSLayoutConstraint.activate([
            timerLabel.topAnchor.constraint(equalTo: view.topAnchor, constant: paddingValue),
            timerLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: paddingValue),
            timerLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -paddingValue),
            timerLabel.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -paddingValue)
        ])
    }

    private func updateLabelStyle() {
        timerLabel.font = UIFont.systemFont(ofSize: fontSize, weight: fontWeight == "bold" ? .bold : .regular)
        timerLabel.textColor = hexStringToUIColor(hexColor: textColor)

        switch textAlign {
        case "center":
            timerLabel.textAlignment = .center
        case "right":
            timerLabel.textAlignment = .right
        default:
            timerLabel.textAlignment = .left
        }
    }

    private func updateViewStyle() {
        if let bgColor = backgroundColor {
            view.backgroundColor = hexStringToUIColor(hexColor: bgColor)
        }

        if let radius = cornerRadius {
            view.layer.cornerRadius = radius
            view.layer.masksToBounds = true
        }
    }

    private func setEndTimeFromString(_ timeString: String) {
        endDate = dateFormatter.date(from: timeString)
        updateDisplay()
    }

    func updateDisplay() {
        guard let endDate = endDate else {
            timerLabel.text = formatTime(days: 0, hours: 0, minutes: 0, seconds: 0)
            return
        }

        let now = Date()
        let timeInterval = endDate.timeIntervalSince(now)

        if timeInterval <= 0 {
            // Timer expired
            timerLabel.text = formatTime(days: 0, hours: 0, minutes: 0, seconds: 0)
            if !wasExpired {
                wasExpired = true
                onExpired?()
            }
            return
        }

        wasExpired = false

        // Check critical threshold
        let isCritical = timeInterval <= criticalThreshold
        if isCritical && !wasCritical {
            wasCritical = true
            onCritical?()
        } else if !isCritical {
            wasCritical = false
        }

        // Update text color based on critical state
        timerLabel.textColor = isCritical ? hexStringToUIColor(hexColor: criticalColor) : hexStringToUIColor(hexColor: textColor)

        // Calculate time components
        let totalSeconds = Int(timeInterval)
        let days = totalSeconds / 86400
        let hours = (totalSeconds % 86400) / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60

        timerLabel.text = formatTime(days: days, hours: hours, minutes: minutes, seconds: seconds)
    }

    private func formatTime(days: Int, hours: Int, minutes: Int, seconds: Int) -> String {
        switch format {
        case "HH:MM:SS":
            if showDays && days > 0 {
                return String(format: "%d:%02d:%02d:%02d", days, hours, minutes, seconds)
            } else if hideZeroHours && hours == 0 {
                return String(format: "%02d:%02d", minutes, seconds)
            } else {
                return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
            }
        case "H:M:S":
            if showDays && days > 0 {
                return String(format: "%d:%d:%d:%d", days, hours, minutes, seconds)
            } else if hideZeroHours && hours == 0 {
                return String(format: "%d:%d", minutes, seconds)
            } else {
                return String(format: "%d:%d:%d", hours, minutes, seconds)
            }
        case "Hh Mm Ss":
            if showDays && days > 0 {
                return String(format: "%dd %dh %dm %ds", days, hours, minutes, seconds)
            } else if hideZeroHours && hours == 0 {
                return String(format: "%dm %ds", minutes, seconds)
            } else {
                return String(format: "%dh %dm %ds", hours, minutes, seconds)
            }
        case "H hours M minutes":
            if showDays && days > 0 {
                return String(format: "%d days %d hours %d minutes", days, hours, minutes)
            } else if hideZeroHours && hours == 0 {
                return String(format: "%d minutes", minutes)
            } else {
                return String(format: "%d hours %d minutes", hours, minutes)
            }
        default:
            return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
        }
    }

    // MARK: - Timer Methods

    func start() {
        isRunning = true
        isPaused = false
        TimerManager.shared.startTimer()
    }

    func pause() {
        isPaused = true
    }

    func resume() {
        isPaused = false
    }

    func reset() {
        wasExpired = false
        wasCritical = false
        updateDisplay()
    }

    func addSeconds(_ seconds: Double) {
        guard let currentEndDate = endDate else { return }
        endDate = currentEndDate.addingTimeInterval(seconds)
        updateDisplay()
    }

    func setEndTime(_ endTime: String) {
        setEndTimeFromString(endTime)
    }

    func getCurrentTime() -> [String: Any] {
        guard let endDate = endDate else {
            return [
                "days": 0,
                "hours": 0,
                "minutes": 0,
                "seconds": 0,
                "totalSeconds": 0,
                "isExpired": true,
                "isCritical": false
            ]
        }

        let now = Date()
        let timeInterval = endDate.timeIntervalSince(now)

        if timeInterval <= 0 {
            return [
                "days": 0,
                "hours": 0,
                "minutes": 0,
                "seconds": 0,
                "totalSeconds": 0,
                "isExpired": true,
                "isCritical": false
            ]
        }

        let totalSeconds = Int(timeInterval)
        let days = totalSeconds / 86400
        let hours = (totalSeconds % 86400) / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60

        return [
            "days": days,
            "hours": hours,
            "minutes": minutes,
            "seconds": seconds,
            "totalSeconds": totalSeconds,
            "isExpired": false,
            "isCritical": timeInterval <= criticalThreshold
        ]
    }

    private func hexStringToUIColor(hexColor: String) -> UIColor {
        let stringScanner = Scanner(string: hexColor)

        if(hexColor.hasPrefix("#")) {
            stringScanner.scanLocation = 1
        }
        var color: UInt32 = 0
        stringScanner.scanHexInt32(&color)

        let r = CGFloat(Int(color >> 16) & 0x000000FF)
        let g = CGFloat(Int(color >> 8) & 0x000000FF)
        let b = CGFloat(Int(color) & 0x000000FF)

        return UIColor(red: r / 255.0, green: g / 255.0, blue: b / 255.0, alpha: 1)
    }
}

// MARK: - Timer Manager Singleton

class TimerManager {
    static let shared = TimerManager()

    private var displayLink: CADisplayLink?
    private var timers: [HybridNitroTimer] = []
    private var isAppInBackground = false

    private init() {
        setupAppStateNotifications()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    private func setupAppStateNotifications() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidEnterBackground),
            name: UIApplication.didEnterBackgroundNotification,
            object: nil
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appWillEnterForeground),
            name: UIApplication.willEnterForegroundNotification,
            object: nil
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
    }

    @objc private func appDidEnterBackground() {
        isAppInBackground = true
        // Keep running timers in background if continueInBackground is true
        for timer in timers {
            if !timer.continueInBackground {
                timer.pause()
            }
        }
    }

    @objc private func appWillEnterForeground() {
        isAppInBackground = false
        // Resume timers that were paused due to backgrounding
        for timer in timers {
            if !timer.continueInBackground && timer.autoStart {
                timer.resume()
            }
        }
    }

    @objc private func appDidBecomeActive() {
        // Force update all timers when app becomes active
        // This ensures accuracy after backgrounding
        DispatchQueue.main.async {
            for timer in self.timers {
                timer.updateDisplay()
            }
        }
    }

    func addTimer(_ timer: HybridNitroTimer) {
        timers.append(timer)
    }

    func removeTimer(_ timer: HybridNitroTimer) {
        timers.removeAll { $0 === timer }
        if timers.isEmpty {
            stopDisplayLink()
        }
    }

    func startTimer() {
        if displayLink == nil {
            setupDisplayLink()
        }
    }

    private func setupDisplayLink() {
        displayLink = CADisplayLink(target: self, selector: #selector(updateTimers))
        displayLink?.add(to: .main, forMode: .common)
    }

    private func stopDisplayLink() {
        displayLink?.invalidate()
        displayLink = nil
    }

    @objc private func updateTimers() {
        for timer in timers {
            let shouldUpdate = timer.isRunning && !timer.isPaused

            // Handle visibility-based pausing
            if timer.pauseWhenNotVisible {
                // In a real implementation, you'd check if the timer's view is visible
                // For now, we'll assume timers are visible when app is in foreground
                let isVisible = !isAppInBackground
                if !isVisible && shouldUpdate {
                    continue // Skip update for invisible timers
                }
            }

            if shouldUpdate {
                DispatchQueue.main.async {
                    timer.updateDisplay()
                }
            }
        }
    }
}
